src/app/upgrade.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2019  Mike Tzou (Chocobo1)
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "upgrade.h"
+
+#include <QFile>
+
+#include "base/profile.h"
+#include "base/settingsstorage.h"
+#include "base/utils/fs.h"
+
+void exportWebUIHttpsFiles();
+
+bool upgrade(const bool /*ask*/)
+{
+    exportWebUIHttpsFiles();
+    return true;
+}
+
+void exportWebUIHttpsFiles()
+{
+    const auto migrate = [](const QString &oldKey, const QString &newKey, const QString &savePath) {
+        SettingsStorage *settingsStorage {SettingsStorage::instance()};
+        const QByteArray oldData {settingsStorage->loadValue(oldKey).toByteArray()};
+        const QString newData {settingsStorage->loadValue(newKey).toString()};
+
+        if (!newData.isEmpty() || oldData.isEmpty())
+            return;
+
+        QFile file(savePath);
+        if (!file.open(QIODevice::WriteOnly))
+            return;
```

Comment:
You just ignore error cases. Isn't there a better way?

Reply:
I'm particular bad at error recovery, is there a better way? I hope it is not too excessive.

Reply:
qBittorrent is designed in such a way that the ability to read/write some folders (i.e. config/data) is its requirement. If there are any problems with this, it cannot function correctly. In any such case, it must interrupt its execution. Console/Log message should be produced.

Reply:
Now I understand your concern.
However in this case, I think it is just simple migration, I don't think failing it is severe enough to stop execution.
Will add log msg for failing cases.

Reply:
>However in this case, it is just simple migration, I don't think failing it is severe enough to stop execution.

I see no difference with the other cases. Really the case is the same. Just ask yourself, what could possibly cause an error here?
>Will add log msg for failing cases.

It is necessary, but it is not enough. Try looking at it from a user perspective. qBittorrent will behave in the same way as without any migration code. The user will not be able to access Web UI for non-obvious reasons, because the application will work (in addition, access via http will be open, which is completely undesirable from a security point of view). Another thing is if the application can not start. This should prompt the user to find out the reason.

Reply:
>in addition, access via http will be open, which is completely undesirable from a security point of view

By the way, this also applies to the current behavior if certificates/keys cannot be loaded. Once a user has configured a secure connection, the application should not allow insecure access for any reason, right? @Piccirello, what do you think?

Reply:
>Really the case is the same. Just ask yourself, what could possibly cause an error here?

Read-only config folder, but I still don't think it should halt program when failed to migrate configs. And there is already log msgs.

>Try looking at it from a user perspective. qBittorrent will behave in the same way as without any migration code. The user will not be able to access Web UI for non-obvious reasons, because the application will work 

If it is the "read-only config folder" case, then qbt will not be working correctly with/without the upgrade...

>in addition, access via http will be open, which is completely undesirable from a security point of view

Could you explain the *security* issue? I'm not aware of it.

>Another thing is if the application can not start. This should prompt the user to find out the reason.

Due to WebUI upgrade code? I must be crazy reading this.
Apparently there are different viewpoints of how important is the upgrade code... @sledgehammer999 your thoughts?

Reply:
>Could you explain the security issue? I'm not aware of it.

For what purpose does the user set up HTTPS? Does it forbid an access via non-secure HTTP? Or am I wrong?
Really I'm not a Web UI user. So I can be wrong about some things. I just express the doubts that arise when viewed from outside.

>Due to WebUI upgrade code? I must be crazy reading this.

Not. But due to invalid execution environment. As I said before qBittorrent is designed assuming the config/data folders is readable/writable. It does not matter when it detects the impossibility of fulfilling this requirement. 

Sorry, nothing personal. It's just that your migration code touched on a fundamental qBittorrent problem that I've been thinking about lately. Inability to read/write config/data files means inability to keep running correctly, so qBittorrent MUST terminate once it faces such problem.

Well, I've made my point, and you decide whether it matters or not.

Reply:
>It's just that your migration code touched on a fundamental qBittorrent problem that I've been thinking about lately.

It is not my concern how "migration code errors" or "read-only config folder" should be handled, it is non-trivial topic itself.
But requiring PR authors in a not-so-related topic to fulfill the blueprints in your mind is just not fair and not justified, this behavior not only slows down the development momentum but also wears down the passion to contribute, in the worst case, throwing away invested efforts.
If you must insist that every PR should provide a fully complete solution to everything it touches, then at least you should be involved in it, such as filling in the things (with exact code) that are not within the PR topic, just throwing expectations and requirements for others to fulfill is not enough.

Reply:
>But requiring PR authors in a not-so-related topic to fulfill the blueprints in your mind is just not fair and not justified, this behavior not only slows down the development momentum but also wears down the passion to contribute, in the worst case, throwing away invested efforts.

That is why I do not insist on my vision of the global problem. If it doesn't look serious to others, just report this fact and keep ignoring it.

@Chocobo1, I'm sorry if my thoroughness offends you. That wasn't my goal.

Reply:
@glassez:  I think `@Chocobo1` was bothered more by your "backseat driving" than your thoroughness.